internal/fs/read_cache_test.go

```diff
@@ -0,0 +1,363 @@
+	err := t.createObjects(objects)
+	AssertEq(nil, err)
+	filePath := path.Join(mntDir, DefaultObjectName)
+	file, err := os.OpenFile(filePath, os.O_RDWR|syscall.O_DIRECT, util.DefaultFilePerm)
+	defer closeFile(file)
+	AssertEq(nil, err)
+
+	// reading object with cache enabled should cache the object into file.
+	buf := make([]byte, len(objectContent))
+	_, err = file.Read(buf)
+	AssertEq(nil, err)
+	AssertEq(objectContent, string(buf))
+
+	objectPath := util.GetObjectPath(bucket.Name(), DefaultObjectName)
+	downloadPath := util.GetDownloadPath(CacheLocation, objectPath)
+	cachedContent, err := os.ReadFile(downloadPath)
+	AssertEq(nil, err)
+	AssertTrue(reflect.DeepEqual(buf, cachedContent))
+}
+
+func cacheFilePermissionWithoutAllowOther(t *fsTest, fileMode os.FileMode) {
+	objectContent := generateRandomString(DefaultObjectSizeInMb * util.MiB)
+	objects := map[string]string{DefaultObjectName: objectContent}
+	err := t.createObjects(objects)
+	AssertEq(nil, err)
+	filePath := path.Join(mntDir, DefaultObjectName)
+	file, err := os.OpenFile(filePath, os.O_RDWR|syscall.O_DIRECT, util.DefaultFilePerm)
+	defer closeFile(file)
+	AssertEq(nil, err)
+
+	// reading object with cache enabled should cache the object into file.
+	buf := make([]byte, len(objectContent))
+	_, err = file.Read(buf)
+	AssertEq(nil, err)
+	AssertEq(string(buf), objectContent)
+
+	objectPath := util.GetObjectPath(bucket.Name(), DefaultObjectName)
+	downloadPath := util.GetDownloadPath(CacheLocation, objectPath)
+	stat, err := os.Stat(downloadPath)
+	AssertEq(nil, err)
+	// confirm file mode is as expected
+	AssertEq(fileMode, stat.Mode())
+}
+
+func writeShouldNotPopulateCache(t *fsTest) {
+	objectContent := generateRandomString(DefaultObjectSizeInMb * util.MiB)
+	filePath := path.Join(mntDir, DefaultObjectName)
+	file, err := os.OpenFile(filePath, os.O_RDWR|syscall.O_DIRECT|os.O_CREATE, util.DefaultFilePerm)
+	defer closeFile(file)
+	AssertEq(nil, err)
+
+	// writing file with cache enabled should not populate cache.
+	buf := []byte(objectContent)
+	n, err := file.Write(buf)
+	AssertEq(nil, err)
+	AssertEq(n, len(objectContent))
+
+	objectPath := util.GetObjectPath(bucket.Name(), DefaultObjectName)
+	downloadPath := util.GetDownloadPath(CacheLocation, objectPath)
+	_, err = os.Stat(downloadPath)
+	AssertNe(nil, err)
+	AssertTrue(strings.Contains(err.Error(), "no such file or directory"))
+}
+
+func sequentialToRandomReadShouldPopulateCache(t *fsTest) {
+	objectContent := generateRandomString(DefaultObjectSizeInMb * util.MiB)
+	objects := map[string]string{DefaultObjectName: objectContent}
+	err := t.createObjects(objects)
+	AssertEq(nil, err)
+	filePath := path.Join(mntDir, DefaultObjectName)
+	file, err := os.OpenFile(filePath, os.O_RDWR|syscall.O_DIRECT, util.DefaultFilePerm)
+	defer closeFile(file)
+	AssertEq(nil, err)
+	// Sequential read
+	buf := make([]byte, util.MiB)
+	_, err = file.Read(buf)
+	AssertEq(nil, err)
+	AssertEq(objectContent[:util.MiB], string(buf))
```

Comment:
While comparing the content, somewhere we have used `AssertEq` and somewhere `AssertTrue with reflect`.

Is there any specific reason? If not we can be consistent while comparing.

Reply:
Changed to AssertTrue with reflect at all places. Reason: The strings in comparison are in MiB and using AssertEq will print the whole strings in case of violation of assertion which is anyways hard to debug.